packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -153,7 +183,15 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>)
     } else {
       lastSelectedKey.current = null;
       setSelectedKey(null);
-      shouldClose = selectedKey === null;
+
+      // If previous key was already null, need to manually call onSelectionChange since it won't be triggered by a setSelectedKey call
+      // This allows the application to control whether or not to close the menu on custom value commit
+      if (selectedKey === null && props.onSelectionChange) {
+        props.onSelectionChange(null);
+      }
+
+      // Should close menu ourselves if component open state or selected key is uncontrolled and therefore won't be closed by a user defined event handler
+      shouldClose = props.isOpen == null || props.selectedKey === undefined;
```

Comment:
Try removing props.isOpen == null here and see if the menu closes when user goes from no selected key to still no selected key via
Combobox (open state is not controlled, selected key is controlled and thus has a onSelectionChange handler)
1. type in field, menu opens
2. hit enter to submit custom value
3. menu closes?

Reply:
Yep, did some testing again, still need the isOpen here